src/coreclr/vm/amd64/thunktemplates.asm

```diff
@@ -0,0 +1,80 @@
+; Licensed to the .NET Foundation under one or more agreements.
+; The .NET Foundation licenses this file to you under the MIT license.
+
+include <AsmMacros.inc>
+include AsmConstants.inc
+
+PAGE_SIZE = 4096
+
+DATA_SLOT macro stub, field
+    exitm @CatStr(stub, <Code + PAGE_SIZE + >, stub, <Data__>, field)
+endm
+
+LEAF_ENTRY StubPrecodeCode, _TEXT
+        mov    r10, QWORD PTR [DATA_SLOT(StubPrecode, MethodDesc)]
+        jmp    QWORD PTR [DATA_SLOT(StubPrecode, Target)]
+LEAF_END_MARKED StubPrecodeCode, _TEXT
+
+LEAF_ENTRY FixupPrecodeCode, _TEXT
+        jmp QWORD PTR [DATA_SLOT(FixupPrecode, Target)]
+PATCH_LABEL FixupPrecodeCode_Fixup
+        mov    r10, QWORD PTR [DATA_SLOT(FixupPrecode, MethodDesc)]
+        jmp    QWORD PTR [DATA_SLOT(FixupPrecode, PrecodeFixupThunk)]
+LEAF_END_MARKED FixupPrecodeCode, _TEXT
+
+LEAF_ENTRY CallCountingStubCode, _TEXT
+        mov    rax,QWORD PTR [DATA_SLOT(CallCountingStub, RemainingCallCountCell)]
+        dec    WORD PTR [rax]
+        je     CountReachedZero
+        jmp    QWORD PTR [DATA_SLOT(CallCountingStub, TargetForMethod)]
+    CountReachedZero:
+        call   QWORD PTR [DATA_SLOT(CallCountingStub, TargetForThresholdReached)]
+LEAF_END_MARKED CallCountingStubCode, _TEXT
+
+LEAF_ENTRY LookupStubCode, _TEXT
+       push   QWORD PTR [DATA_SLOT(LookupStub, DispatchToken)]
+       jmp    QWORD PTR [DATA_SLOT(LookupStub, ResolveWorkerTarget)]
+LEAF_END_MARKED LookupStubCode, _TEXT
+
+LEAF_ENTRY DispatchStubCode, _TEXT
```

Comment:
I am surprised that you were able to get away with the extra indirection in DispatchStubCode. It will be interesting to see the results of the microbenchmark runs.

Reply:
I am not sure what indirection you mean. This was the original code:
https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/coreclr/vm/amd64/virtualcallstubcpu.hpp#L286-L289
https://github.com/dotnet/runtime/blob/a1a653c7217706fcb820914d65126234bc2447a8/src/coreclr/vm/amd64/virtualcallstubcpu.hpp#L183-L191

Reply:
I meant the performance overhead from the extra indirections introduced by this refactoring. The original code had two immediates, the new code has two indirections.  

@adamsitnik What's the best way to do dotnet/performance benchmark run for a PR like this one to see what improved/regressed on different target platforms?

Reply:
> What's the best way to do dotnet/performance benchmark run for a PR like this one to see what improved/regressed on different target platforms?

We have two options:

1. Run the benchmarks using two different coreruns (before & after) and store them in dedicated folders, then compare them using ResultsComparer tool: https://github.com/dotnet/performance/blob/ef3edbc52d92c6b30ba1c316082f46865e5ff1d6/docs/benchmarking-workflow-dotnet-runtime.md#preventing-regressions It's the best solution if the implementation might change and you might need to re-run the benchmarks using updated corerun.
2. Run the benchmarks using two different coreruns without storing them in dedicated dolders, use BDN to perform the statistical test on the fly: https://github.com/dotnet/performance/blob/03207f183f042f6fc6b9f341df7a0e36b7175f5d/src/benchmarks/micro/README.md#private-runtime-builds it's good if you have only few benchmarks or don't intend to change the implementation